File: src/main/java/testing/EmbeddedServerHelper.java
package testing;

import logging.LogHelper;
import org.neo4j.driver.v1.Driver;
import org.neo4j.driver.v1.GraphDatabase;
import org.neo4j.driver.v1.Session;
import org.neo4j.graphdb.GraphDatabaseService;
import org.neo4j.harness.ServerControls;
import org.neo4j.harness.TestServerBuilders;
import org.slf4j.Logger;

import java.net.URI;

public class EmbeddedServerHelper {

    private static ServerControls embeddedDatabaseServer;
    private static final Logger log = LogHelper.getLogger();
    private static GraphDatabaseService graphDb;

    public static GraphDatabaseService getEmbeddedServer() {
        if (graphDb != null) {
            return graphDb;
        }
        embeddedDatabaseServer = TestServerBuilders
                .newInProcessBuilder()
                .withConfig("dbms.connector.bolt.enabled", "true")
                .withConfig("dbms.connector.bolt.listen_address", ":8888")
                .withConfig("dbms.logs.query.enabled", "true")
                .withConfig("dbms.track_query_cpu_time", "true")
                .newServer();

        log.info(String.format("Bolt server is available at: %s", embeddedDatabaseServer.boltURI()));
        graphDb = embeddedDatabaseServer.graph();
        return graphDb;
    }

    public static URI getBoltUri() {
        return embeddedDatabaseServer.boltURI();
    }

    public static void generateGraphFromJsonTemplate(String jsonFilePath) {
        GraphFromJsonGenerator graphFromJsonGenerator = new GraphFromJsonGenerator(getEmbeddedServer(), jsonFilePath);
        graphFromJsonGenerator.generateGraph();
    }

    public static void clearGraph() {

        try (
                Driver driver = GraphDatabase.driver(getBoltUri());
                Session session = driver.session()
        ) {
            log.debug("Clearing the graph using detach delete...");
            session.run("MATCH (n) detach delete n");
        }

    }

    private EmbeddedServerHelper() {
    }
}

File: src/main/kotlin/utilities/EmbeddedServerHelper.kt
package utilities

import logging.LogHelper
import org.neo4j.driver.v1.GraphDatabase
import org.neo4j.driver.v1.StatementResult
import org.neo4j.graphdb.GraphDatabaseService
import org.neo4j.harness.ServerControls
import org.neo4j.harness.TestServerBuilders
import java.net.URI

object EmbeddedServerHelper {
    private val embeddedDatabaseServer: ServerControls = TestServerBuilders
            .newInProcessBuilder()
            .withConfig("dbms.connector.bolt.enabled", "true")
            .withConfig("dbms.connector.bolt.listen_address", ":8888")
            .withConfig("dbms.logs.query.enabled", "true")
            .withConfig("dbms.track_query_cpu_time", "true")
            .newServer()
    private val log = LogHelper.logger
    val graphDb: GraphDatabaseService
    private val boltUri: URI

    init {

        graphDb = embeddedDatabaseServer.graph()
        boltUri = embeddedDatabaseServer.boltURI()
        log.info("Bolt server is available at: $boltUri")
    }

    fun execute(query: String): StatementResult {
        GraphDatabase.driver(boltUri).use { driver ->
            driver.session().use { session ->
                log.debug("Executing: $query")
                return session.run(query)
            }
        }
    }

    fun clearGraph() {

        GraphDatabase.driver(boltUri).use { driver ->
            driver.session().use { session ->
                log.debug("Clearing the graph using detach delete...")
                val run = session.run("MATCH (n) detach delete n")
                run
            }
        }

    }
}
